### ISSUE TITLE
Web Certificate not being used

### ISSUE THREAD

**Issue by dkowis** [goauthentik/authentik#7173]
**Describe the bug**
I've configured the default tenant to use a web certificate.
![image](https://github.com/goauthentik/authentik/assets/28048/2edb7b01-7176-4c62-a1d6-c3dfc89e7017)

According to [the documentation](https://goauthentik.io/docs/core/certificates#web-certificates) I've done everything correctly, and it should be serving the webui using that certificate.

I found a [duplicate issue](https://github.com/goauthentik/authentik/issues/4182), but the "answer" there was that after a few hours it suddenly started working. Sadly, mine has not resolved itself. After several hours, it is still serving the self-signed certificate, instead of the one I've selected.

**To Reproduce**

1. Add certs to the container. I used prepopulated certs, which were then imported when the task ran.
2. Update the default tenant to use that web certificate.
3. 

**Expected behavior**
When inspecting the cert I should see the one I provided, signed by my internal CA, not the self-signed cert.

**Screenshots**
Here, I have the default tenant updated to use the web certificate specified, and yet I'm still being served the default certificate.
![image](https://github.com/goauthentik/authentik/assets/28048/7531f1cb-fa99-4ee8-8f32-a711f24e223f)

Here's the certificate itself in authentik's UI
![image](https://github.com/goauthentik/authentik/assets/28048/06de8959-e99c-4f35-b7bb-408c633b5a10)

**Logs**
Here's the log from the server and worker
- [log.zip](https://github.com/goauthentik/authentik/files/12908396/log.zip)
- [worker_log.zip](https://github.com/goauthentik/authentik/files/12908397/worker_log.zip)

**Version and Deployment (please complete the following information):**

-   authentik version: 2023.8.3
-   Deployment: docker-compose

**Additional context**
Literally first time install, I'm trying to get a ldap outpost to work too, which isn't being healthy, and I think maybe because of the certificates, but I'm not sure there. (I think that's the source of all the `TLS handshake error from 10.200.0.1:39626: remote error: tls: bad certificate` log entries.

**Comment by dkowis** [comment]
Oh heck. I'm discovering that it's doing some SNI things somehow. 

When I access it via IP address, I get the default certificate, when I access it via a domain name, I get the certificate, which isn't valid, because it doesn't include the port, which is fine, and I can fix that.

I expected it to use the certificate anyway, since I have a Subject Alternative Name configured for the IP address so that it would work. Perhaps it's some other SAN matching? I'm going to try something.

**Comment by dkowis** [comment]
Okay, Yeah this is what's going on, and it was super unexpected.

When I hit authentik direct via IP, it gives me the self-signed cert, not the one I have assigned to the default tenant. 

When I hit it via DNS name, it serves the correct certificate.

Is there a way to ensure it always gives me the certificate, regardless of IP or DNS name when accessing the server?

**Comment by hotrod225** [comment]
I'm seeing this same behavior myself. Actually caused me to spend alot of wasted time troubleshooting certs while connecting authentik to my vcenter

**Comment by dkowis** [comment]
Well it is still a problem... so I guess this will bump it alive

**Comment by sudoshi** [comment]
Bump.  Same problem.

**Comment by Luccifer** [comment]
Yep, got the same problem in nightly builds of MailCow at the attempt of OIDC.. Even added SSLs from nginx-reverse-proxy to tenants, still 
`http: TLS handshake error from 192.168.88.1:49016: remote error: tls: unknown certificate authority`

**Comment by gmpreussner** [comment]
What worked for me was to shut down and restart the containers, after which the default tenant / admin UI used the selected certificate correctly:
```
$ sudo docker compose down
$ sudo docker compose up -d
```

That being said, this only worked for me with a manually imported certbot certificate pair. Automatic certificate detection via the `/cert` folder seems to be completely broken; and it's not a problem with the volume mappings - they show up inside the container with the correct permissions and file contents.

See also:
- https://github.com/goauthentik/authentik/issues/5888
- https://github.com/goauthentik/authentik/issues/7984

**Comment by sudoshi** [comment]
Is anyone going to pay attention to this issue?  Bump.

**Comment by dasrfp** [comment]
it's not "stale", it's being ignored. Same problem here

**Comment by joelmccoy** [comment]
I am having this same issue when deploying this to kubernetes with the helm chart.  I have a cert setup on the ingress but I am getting an internal server error. 
![image](https://github.com/goauthentik/authentik/assets/14078350/8deb2854-1e19-46fe-a5a2-f914b6e16e62)
The pod logs show: `remote error: tls: bad certificate`

**Comment by dromix** [comment]
@joelmccoy Hi, did you fix it somehow? I've met it recently too

**Comment by Jessychan98** [comment]
Hi, i added the following at the end of the docker compose file and the certificate in the /certs are getting used.
worker:
  driver: local

![Screenshot 2024-05-07 130435](https://github.com/goauthentik/authentik/assets/72158996/e1d58dde-b4c8-4a4f-993e-0ba12fcebbba)

**Comment by los93sol** [comment]
@BeryJu Is there a plan to fix certificate_discovery for k8s setups?  This is a huge broken and the import_certificate command with a cron is hardly a user friendly solution.  This has been broken for a VERY long time now with little effort to resolve it

**Comment by BeryJu** [comment]
> @BeryJu Is there a plan to fix certificate_discovery for k8s setups?  This is a huge broken and the import_certificate command with a cron is hardly a user friendly solution.  This has been broken for a VERY long time now with little effort to resolve it

certificate_discovery should work for k8s just as it does for compose, there has been a bug more recently that scheduled tasks don't run however that's fixed with 2024.4. If there's another issue with it I'm not directly aware of it.

**Comment by los93sol** [comment]
Unfortunately, it does not work as expected.  I'm using the helm chart with....

server:
  volumeMounts:
    - name: ldap-tls
      mountPath: /certs/ldap.example.com
    - name: authentik-tls
      mountPath: /certs/auth.example.com.pem
      subPath: tls.crt
      readOnly: true
    - name: authentik-tls
      mountPath: /certs/auth.example.com.key
      subPath: tls.key
      readOnly: true

  volumes:
    - name: ldap-tls
      secret:
        secretName: ldap-tls
        items:
          - key: tls.crt
            path: fullchain.pem
          - key: tls.key
            path: privkey.pem
    - name: authentik-tls
      secret:
        secretName: authentik-tls

I can see the certs in the pod and verified their content is valid, but neither method of mounting works.
Manually running certificate_discovery show 0 imported, but when I use import_certificate it does import it properly so something definitely seems to not be working as expected.

**Comment by los93sol** [comment]
Unfortunately there's no errors showing so I'm totally confused, but a quick scan through issues led me to several users having issues with this.

https://github.com/goauthentik/authentik/commit/8b95e9f97ac7d2e8860fec4291e2237c21dc906c is the commit with the read only open fix I was talking about

**Comment by los93sol** [comment]
Trying to debug this locally, I am attempting to overwrite /authentik/crypto/tasks.py with a version of it in a configmap, but it still seems to be executing the original version.  Is there a way to change it's content without building authentik completely?

**Comment by los93sol** [comment]
I ended up getting it to work, turns out you have to mount the volumes to the worker, not to the server.  When I was running the import command I was doing it on the server which is why it worked but not the discovery

**Comment by BeryJu** [comment]
To answer the 3 different issues in this thread:
- Access via IP: authentik will try to find the correct certificate for a brand based on the DNS name the request comes from and what is configured in the brand, however it does not currently respect the `Default` flag and will as such fall back to a self-signed certificated; this is a bug and will be fixed
- Use with an ingress: When running authentik in Kubernetes, it is generally recommended to use ingresses and _not_ import the HTTPS certificate into authentik as that part is done by the ingress
- Certificate import: the certificates indeed need to be mounted in the worker, which is mentioned in the docs for the import command/automated import process

**Comment by Torxed** [comment]
> Certificate import: the certificates indeed need to be mounted in the worker, which is mentioned in the docs for the import command/automated import process: https://docs.goauthentik.io/docs/core/certificates#manual-imports

Maybe I misunderstand the thread here. But even with a (successful) manual import, it's not being used for the web service if the certificate is a wildcard certificate.

In our case we have a certificate `*.domain.com` that gets imported as `Name: website` and is not being used. If we leave it up to auto-discovery it doesn't find the private key (not sure if it's because it's a EC key or because it's going by filename and is expecting a very specific format. In our case it's `wildcard_.domain.com.certificate.pem` + `wildcard_.domain.com.privkey.pem`.

Either way, how do you get the web service to use a specific given certificate?

**Comment by anuneo** [comment]
> To answer the 3 different issues in this thread:
> 
> * Access via IP: authentik will try to find the correct certificate for a brand based on the DNS name the request comes from and what is configured in the brand, ~however it does not currently respect the `Default` flag and will as such fall back to a self-signed certificated; this is a bug and will be fixed~
>   re-reviewing the code, authentik does consider the default flag. Since the domain of the brand was changed I'm assuming the default blueprint ran and created a new default brand which was used as a fallback in this case
> * Use with an ingress: When running authentik in Kubernetes, it is generally recommended to use ingresses and _not_ import the HTTPS certificate into authentik as that part is done by the ingress
> * Certificate import: the certificates indeed need to be mounted in the worker, which is mentioned in the docs for the import command/automated import process: https://docs.goauthentik.io/docs/core/certificates#manual-imports

Hi @BeryJu,
I am facing this problem as well when I try to access via IP. I just have only one brand where I have configured to use the desired certificate. Still it is not getting used.

**Comment by umer936** [comment]
I was having trouble with this for a while, but then got it. 

I did 

> Admin > System > Brands 

Domain: systemname.url.tld (my cert is *.url.tld) [Example]

> Other global settings

Web Certificate: select your cert (I kept missing this option!) 
Selected certs: selected it here


then 
`docker compose down`
`docker compose up -d` 


--

Not sure if that solves the things discussed here, but it got me past where I was stuck.

**Comment by crazyyzarc** [comment]
> @BeryJu 
> To answer the 3 different issues in this thread:
> 
>     * Access via IP: authentik will try to find the correct certificate for a brand based on the DNS name the request comes from and what is configured in the brand, ~however it does not currently respect the `Default` flag and will as such fall back to a self-signed certificated; this is a bug and will be fixed~
>       re-reviewing the code, authentik does consider the default flag. Since the domain of the brand was changed I'm assuming the default blueprint ran and created a new default brand which was used as a fallback in this case

All of our services also have the static IP stored in the certificate. And despite its availability, Authentik still uses the Authentik Default Certificate. Why? As users, can we activate IP in certificate compatibility via a flow?

**Comment by NielsDa** [comment]
I have a similar issue, I'm new to Authentik, I have just installed 2025.10.0

I Name and paste the LetsEncrypt certificate and key into the _System->Certificates>Import_ UI.
In  _System->Brand>Edit_ I change the selection from the selfsigned certificate, to the name given above.

When I connect via. https i get a security warning in my browser since it still gets the self-signed certificate.

If I delete the self-signed certificate in   _System->Certificates_  then I get the LetsEncrypt certificate as expected.
If I restart the docker container, it generates a new selfsigned certificate, and used this for https.

So after every restart I have to accept the unsafe certificate in my browser, and then log in and delete the newly selfsigned certificate, othervise https fails.

Question: 
Are the Authentik running its own DNS resolver ? and trying to be "smart" about the certificate selection?
I'm running an split DNS on my domain, so the DNS lookup will give different IP depending if you ask the WAN side or the LAN side DNS. IP address of LAN side DNS is server is provided by DHCP,

**Comment by kub3let** [comment]
Same here Authentik ignores the global certificate set and uses sni inspection instead.

I ended up setting host overrides and using hostname instead of ip as a workaround.

But then there is also the point that there is no support for setting the certificate on the outpost proxy ? 

I don't want to end up with manual certificate pinning...

Leaves a really bad taste that Authentik is so open to MITM attacks.

---

With nginx you can also override the SNI send to force it.

```
# force SNI
proxy_ssl_name authentik-server;
proxy_ssl_server_name   on;
```

**Comment by Frill7742** [comment]
I have the same issue, and I'm unable to fix it.

I would like to move from the default and insecure RSA 2046 to a more modern, faster and more secure system like ED25519, however due to this bug I'm unable to do so.
@BeryJu   and to be honest with the authentik team here it leaves a terrible after-taste for a software suite that is so critical and that adversities with being secure to force RSA 2046.

This GitHub issue is here for over two years now and the last response after it was closed did not address the issue. Since the fix actually does not fix anything at all, since the mentioned certificate import does not work.

Edit:
I tried it with ECDSA, ED25519 and ED448. I tried it with Certs that are generated for the IPv4 address where the SNI should be the same (it is behind a Caddy reverse proxy sitting on another Server) and I tried all the different certs with the actual domain. It also fails.

I could confirm that the host, caddy and the openSSL version of the reverse proxy supports ED25519 and ECDSA. I did not verify if ED448 was supported by the client.

**Comment by bkoragan** [comment]
@Frill7742 Here is the PR # https://github.com/goauthentik/authentik/pull/21417 for fixing this reported issue. PR has been accepted & merged, and cheery picked this fix to **version-2026.2.**

**Comment by Torxed** [comment]
> [@Frill7742](https://github.com/Frill7742) Here is the PR # [#21417](https://github.com/goauthentik/authentik/pull/21417) for fixing this reported issue. PR has been accepted & merged, and cheery picked this fix to **version-2026.2.**

Well done! Three years of silence — gave up on this ever being fixed. I'm genuinely impressed and will re-evaluate authentik, thank you! :)
